### PR TITLE
Update memory tracker periodically with cgroup memory usage

### PR DESCRIPTION
**Pull request by rschu1ze**
When the system runs out of memory (as determined by system-wide settings or cgroups memory thresholds), the OOM killer stops memory-intensive processes. Due to its nature, ClickHouse will often be the victim. To avoid this, the upper limit of the (global) memory tracker in ClickHouse can be configured to a value smaller than the "kill threshold", see setting [max_server_memory_usage_ram_ratio](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#max_server_memory_usage_to_ram_ratio).

There are some problems with this approach:

- The maximum memory threshold of the processes's cgroup can be changed (increased, decreased) while ClickHouse runs. An earlier fix (#53455) ensured that new memory limits were at least considered when ClickHouse's configuration changes (a follow-up fix #59435 resolved problems with nested cgroups). Of course, users need to make real or dummy configuration changes for this to happen and it may already be too late at this point.

- The memory tracking in ClickHouse is not 100% precise. For example, ClickHouse may think it uses *less* memory than the OS thinks. In that case, there is a risk of getting OOM-killed. Or the other way round, ClickHouse produces OOMs even if there is sufficient memory.

This PR tries to address this by reading the current memory consumption as seen by cgroups periodically and running two possible actions:
- if actual memory consumption (as read from cgroups) exceeds a soft threshold, purge jemalloc preallocated arenas. This should bring the OS's and ClickHouse's view of the memory consumption closer together.
- if a hard threshold is exceeded, adjust (usually lower) the hard limit in the memory tracker to the exceeded threshold. This increases the chance for OOMs in ClickHouse but decreases the risk of getting OOM killed.

The read interval (setting `cgroups_memory_usage_observer_wait_time`) is 15 sec by default and it can be reduced to 1 sec if needed. 0 means disabled. The parameter cannot be changed after database start.

The hard and soft limits are controlled by parameters `cgroup_memory_watcher_hard_limit_ratio` and `cgroup_memory_watcher_soft_limit_ratio` (defaults: 0.95, 0.9). They each refer to the initial maximum allowed global memory consumption computed as `total system memory x max_server_memory_usage_to_ram_ratio`. Both parameters can be changed at runtime.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by rschu1ze** [comment]
Note to myself: needs similar changes in #59435

File: src/Common/CgroupsMemoryUsageObserver.cpp [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+/// - All of the logic in this file assumes that the current process is the only process in the
+///   containing cgroup (or more precisely: the only process with significant memory consumption).
+///   If this is not the case, then other processe's memory consumption may affect the internal
+///   memory tracker ...
+/// - Cgroups v1 and v2 allow nested cgroup hierarchies. As v1 is deprecated for over half a
+///   decade and will go away at some point, hierarchical detection is only implemented for v2.
+/// - I did not test what happens if a host has v1 and v2 simultaneously enabled. I believe such
+///   systems existed only for a short transition period.
+
+std::optional<std::string> getCgroupsV2FileName()
```
Comment (hanfei1991) [inline src/Common/CgroupsMemoryUsageObserver.cpp:123]: there are similar logic in `base/getMemoryAmount` to read V2File. The difference seems to be getting different sections of the file. Is it possible to reuse some code?

Comment (rschu1ze) [reply src/Common/CgroupsMemoryUsageObserver.cpp:123]: Will do that, yes.

File: src/Common/CgroupsMemoryUsageObserver.cpp [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        {
+            /// Format is
+            ///   kernel 5
+            ///   rss 15
+            ///   [...]
+            std::string key;
+            while (!buf.eof())
+            {
+                readStringUntilWhitespace(key, buf);
+                if (key != "rss")
```
Comment (alesapin) [inline src/Common/CgroupsMemoryUsageObserver.cpp:220]:
```suggestion
                if (key != "rss") 
                {
                    std::string dummy;
                    readStringUntilNewlineInto(dummy, buf);
                    continue;
                }
```

File: src/Common/CgroupsMemoryUsageObserver.cpp [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                    continue;
```
Comment (alesapin) [inline src/Common/CgroupsMemoryUsageObserver.cpp:221]:
```suggestion
```

**Pull request merged**